nvdaHelper/remote/displayModel.cpp

```diff
@@ -139,22 +145,43 @@ void displayModel_t::clearAll() {
 	setFocusRect(NULL);
 }
 
-void displayModel_t::clearRectangle(const RECT& rect, BOOL clearForText) {
+void displayModel_t::clearRectangle(const RECT& rect, BOOL clearForText, HDC clipRgnHdc) {
 	LOG_DEBUG(L"Clearing rectangle from "<<rect.left<<L","<<rect.top<<L" to "<<rect.right<<L","<<rect.bottom);
 	set<displayModelChunk_t*> chunksForInsertion;
 	displayModelChunksByPointMap_t::iterator i=chunksByYX.begin();
 	RECT tempRect;
 	//If the rectangle we are clearing completely covers any current focus rectangle, then get rid of the focus rectangle.
-	if(focusRect&&IntersectRect(&tempRect,&rect,focusRect)&&EqualRect(&tempRect,focusRect)) {
+	if(
+		// A focus rectangle is currently set.
+		focusRect
+		// Either we don't check for clipping or the focus rectangle is part of the clipping region.
+		&&!(clipRgnHdc!=NULL&&RectVisible(clipRgnHdc,focusRect)==FALSE)
+		// The focus rectangle falls completely within the rectangle to clear.
+		&&IntersectRect(&tempRect,&rect,focusRect)
+		&&EqualRect(&tempRect,focusRect)
+	) {
 		setFocusRect(NULL);
 	}
+	RECT tempLogicalRect;
 	while(i!=chunksByYX.end()) {
 		displayModelChunksByPointMap_t::iterator nextI=i;
 		++nextI; 
 		displayModelChunk_t* chunk=i->second;
 		int baseline=i->first.first;
 		if(IntersectRect(&tempRect,&rect,&(chunk->rect))) {
 			//The clearing rectangle intercects the chunk's rectangle in some way.
+			if(clipRgnHdc != NULL) {
+				tempLogicalRect = tempRect;
+				dcPointsToScreenPoints(clipRgnHdc,(LPPOINT)&tempLogicalRect,2,false);
+				//RectVisible is very poorly documentated on the web.
+				//It can return -1 for an invalid hdc, which is not documented,
+				//and is able to return several other integers, while we expect a BOOL.
+				if(RectVisible(clipRgnHdc,&tempLogicalRect)==FALSE) {

```

Comment:
VisibleRect takes logical coordinates, but you just converted them above to device coordinates?
Also, I think there might be an error in the windows documentation of rectVisible:
"If the current transform has a rotation and the rectangle does not lie within the clipping region, the return value is 1."
I'm guessing they actually mean it returns -1. Otherwise they are re-using the value 1 for two contridictory situations.

Reply:
Ugh, now I see. this really should be a call of screenToDCPoints instead. I will test this next friday.